Ginkgo update + stack fix

pohly opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

After merging
https://github.com/kubernetes/kubernetes/commit/259bb3bef5403daec1bbe44c2c9413ec17b52b6e#diff-eb7b79470992813ea1905e96c298b47b
ExpectEqual and some of the other wrappers logged the failure twice,
once inside the wrapper itself and once in the failure handler.
    
Logging the stack backtrace is useful because many assertions still
don't contain an explanation and therefore knowing where they occur is
crucial. Now all failures are logged with a "Full Stack Trace", not
just those with a wrapper. The stack is pruned to skip over wrapper
functions and removes Ginkgo internal functions to keep the stack
trace smaller.
    
Failures occurring in the wrappers were recorded as occurring in those
wrappers. Now the wrappers are skipped and the caller is recorded
instead.

Fixes: #82013

**Does this PR introduce a user-facing change?**:
```release-note
more complete and accurate logging of stack backtraces in E2E failures
```

pohly committed:
e2e log: fix full stacktrace with Ginkgo 1.10.0

Ginkgo 1.10.0 includes the relevant fix for dumping the full stack
(https://github.com/onsi/ginkgo/pull/590), so when using that release
we can simplify the logging unit test.

By changing the skipping, we can avoid the rather volatile util.go
entries. However, that gomega is part of the stack trace still needs
to be fixed in Gingko.

pohly committed:
e2e log: skip log.Failf in stack information

The function itself isn't relevant.

pohly committed:
e2e log: consistent logging of stack backtrace and failure, part II

After merging
https://github.com/kubernetes/kubernetes/commit/259bb3bef5403daec1bbe44c2c9413ec17b52b6e#diff-eb7b79470992813ea1905e96c298b47b
ExpectEqual and some of the other wrappers logged the failure twice,
once inside the wrapper itself and once in the failure handler.

Logging the stack backtrace is useful because many assertions still
don't contain an explanation and therefore knowing where they occur is
crucial. Now all failures are logged with a "Full Stack Trace", not
just those with a wrapper. The stack is pruned to skip over wrapper
functions and removes Ginkgo internal functions to keep the stack
trace smaller.

Failures occuring in the wrappers were recorded as occuring in those
wrappers. Now the wrappers are skipped and the caller is recorded
instead.

The full stack trace recorded by Ginkgo 1.10.0 is currently off by one
entry. This needs to be fixed in Ginkgo, then the test can be updated.

pohly commented:
This is WIP because Ginkgo 1.10.0 still isn't quite working correctly: https://github.com/onsi/ginkgo/pull/600

pohly committed:
e2e log: Ginkgo 1.10.1 fixes stack skip

Now Ginkgo properly skips the initial stack entry.

pohly commented:
/retest

oomichi commented on test/e2e/framework/log/logger.go:
```diff
@@ -59,6 +63,42 @@ func Fail(msg string, callerSkip ...int) {
+// entries coming from Ginkgo.
+//
+// This is a copy of https://github.com/onsi/ginkgo/blob/f90f37d87fa6b1dd9625e2b1e83c23ffae3de228/internal/codelocation/code_location.go#L25
+func PrunedStack(skip int) string {
```
The original function name is PruneStack in ginkgo. I guess this name (with `d`) is intentional, right?
(I tried to search this name in ginkgo repo, and I missed that with the name)

pohly replied:
I intentionally renamed it, because this version is simpler and returns the "pruned stack" instead of taking an arbitrary string and pruning it.

pohly replied:
I added "This is a *modified* copy of PruneStack in ..."

oomichi commented:
/cc @oomichi

alejandrox1 commented:
/assign

alejandrox1 commented:
/cc

pohly commented on test/e2e/framework/log/logger.go:
```diff
@@ -19,6 +19,9 @@ package log
 
 import (
```
I see. I've moved it and also some of the `util.go` functions that are only used in `log.go`. I also updated the comment in `framework/e2e/log/logger.go` to be more explicit.

pohly commented on test/e2e/framework/log/logger.go:
```diff
@@ -48,8 +51,9 @@ func Failf(format string, args ...interface{}) {
 // (for example, for call chain f -> g -> FailfWithOffset(1, ...) error would be logged for "f").
```
Done.

pohly committed:
e2e log: move back to framework, part II

This continues the work in
https://github.com/kubernetes/kubernetes/pull/81426 by also moving the
logger_test.go, moving the log helper code from util.go to log.go (a
more logical place, as it is only used there) and updating comments.

pohly commented:
I found that the log unit test stopped working on September 1st, see fix in https://github.com/kubernetes/kubernetes/pull/82229

I'd like to get that merged for master and 1.16 before continuing here.

pohly committed:
bazel update

pohly commented:
@oomichi: PR updated, all tests clean -> please have another look.

oomichi commented on test/e2e/framework/log/logger.go:
```diff
@@ -14,7 +14,8 @@ See the License for the specific language governing permissions and
 
-// Package log should be removed after switching to use core framework log.
+// Package log will be removed after switching to use core framework log.
+// Do not make further changes here!
```
+1 :-)

oomichi reviewed (commented):
@pohly Thank you so much for this work!

/approve
/lgtm

alejandrox1 reviewed (approved):
This looks good!
/milestone v1.17
/priority backlog

pohly commented:
@oomichi: I had to rebase to resolve a conflict in test/e2e/framework/BUILD, can you re-add your lgtm?

alejandrox1 commented:
mmm :thinking: 
/retest

pohly commented:
"mmm" indeed. No output whatsoever from the failed job?

chases2 commented:
Hello, all!

I'm currently looking into why this isn't generating logs. So far I know:
- The job did actually run
- It actually seem to have emitted no logs (likely not a UI issue)

pohly commented:
`make test` works locally, but isn't quite the same (no bazel).

pohly commented:
/retest

oomichi commented:
Thanks again

/lgtm

pohly commented:
`bazel test --config=unit  //... -- -//build/... -//vendor/...` also worked locally.

pohly commented:
... and the third attempt today produces output: https://prow.k8s.io/log?id=1176193418221588482&job=pull-kubernetes-bazel-test

alejandrox1 commented:
Just need a final approval
/assign @oomichi @timothysc @andrewsykim 
ptal

alejandrox1 commented:
Thank you @chases2 for taking a look at the failures

andrewsykim commented on test/e2e/framework/log.go:
```diff
@@ -18,24 +18,88 @@ package framework
+// This is a modified copy of PruneStack in https://github.com/onsi/ginkgo/blob/f90f37d87fa6b1dd9625e2b1e83c23ffae3de228/internal/codelocation/code_location.go#L25:
+// - simplified API and thus renamed (calls debug.Stack() instead of taking a parameter)
+// - source code filtering updated to be specific to Kubernetes
+func PrunedStack(skip int) string {
```
Anyway this can be consolidated with https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/ginkgowrapper/wrapper.go#L112?

pohly replied:
That may be useful, but I have no idea how that code is used and therefore would prefer to not touch it.

oomichi commented:
@alejandrox1 
> Just need a final approval

This PR changes many files then we need approval by approver of root directory.
(I've put my best score on this already :-)

timothysc commented on go.mod:
```diff
@@ -93,7 +93,7 @@ require (
 	github.com/mvdan/xurls v1.1.0
-	github.com/onsi/ginkgo v1.8.0
+	github.com/onsi/ginkgo v1.10.1
 	github.com/onsi/gomega v1.5.0
```
Shouldn't this also be updated to 1.7.0 if you're making this change?

pohly replied:
It could get updated, but it's unrelated to the problem that this PR is solving and both projects are only loosely related, so Ginkgo 1.10.1 and Gomega 1.5.0 should be a valid combination.

Considering that both the Ginkgo and the Gomega update could cause issues, it might be safer to update them separately.

timothysc replied:
The releases we're clearly coordinated when looking at the commit history, please update gomega to ensure we are in lock step.

timothysc replied:
/hold

pohly replied:
> The releases we're clearly coordinated when looking at the commit history, please update gomega to ensure we are in lock step.

Okay, if that's preferred.

pohly replied:
/hold cancel

Gomega updated, too. I also rebased.

pohly committed:
vendor: update gomega to v1.7.0

Updated to keep both Ginkgo and Gomega at the latest releases.

alejandrox1 reviewed (commented):
/lgtm
/retest

/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-area kubectl
/remove-area cloudprovider
/remove-area apiserver

timothysc reviewed (approved):
/approve

timothysc commented:
/assign @liggitt  
^ requires approval for dep changes.

liggitt commented:
/approve

alejandrox1 commented:
@pohly could you try doing a `git commit --amend && git push -f origin ginkgo-stack-fix` to try and remove the needs-rebase label ? :thinking:

cblecker commented:
/test all

pohly commented:
> could you try doing a git commit --amend && git push -f origin ginkgo-stack-fix to try and remove the needs-rebase label ?

Done. @alejandrox1 now you need to LGTM again.

oomichi commented:
/lgtm
